scipy/linalg/_cythonized_array_utils.pyx

```diff
@@ -66,34 +66,23 @@ def find_det_from_lu(lapack_t[:, ::1] a):
 
 
 # ====================== swap_c_and_f_layout : s, d, c, z ====================
-@cython.cdivision(True)
+@cython.nonecheck(False)
 @cython.wraparound(False)
 @cython.boundscheck(False)
 @cython.initializedcheck(False)
-cdef void swap_c_and_f_layout(lapack_t *a, lapack_t *b, int r, int c, int n) noexcept nogil:
-    """Recursive matrix transposition for square arrays"""
-    cdef int i, j, ith_row, r2, c2
-    cdef lapack_t *bb=b
-    cdef lapack_t *aa=a
-    if r < 16:
-        for j in range(c):
-            ith_row = 0
-            for i in range(r):
-            # Basically b[i*n+j] = a[j*n+i] without index math
-                bb[ith_row] = aa[i]
-                ith_row += n
-            aa += n
-            bb += 1
-    else:
-        # If tall
-        if (r > c):
-            r2 = r//2
-            swap_c_and_f_layout(a, b, r2, c, n)
-            swap_c_and_f_layout(a + r2, b+(r2)*n, r-r2, c, n)
-        else:  # Nope
-            c2 = c//2
-            swap_c_and_f_layout(a, b, r, c2, n);
-            swap_c_and_f_layout(a+(c2)*n, b+c2, r, c-c2, n)
+cdef inline void swap_c_and_f_layout(lapack_t *a, lapack_t *b, int r, int c) nogil:
```

Comment:
```suggestion
cdef inline void swap_c_and_f_layout(lapack_t *a, lapack_t *b, int r, int c) noexcept nogil:
```

Just a humble suggestion: Cython 3 is going to change default handling of exceptions - see https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html#exception-values-and-noexcept

Cython 0.29.31 and newer also support `noexcept` keyword as noop. Using `noexcept` keyword can simplify migration to Cython3.

There are already PRs related to this: https://github.com/scipy/scipy/pull/18266 and https://github.com/scipy/scipy/pull/18362